engine: add histogram_quantile test case with malformed data

GiedriusS opened the pull request.

GiedriusS commented on execution/function/histogram.go:
```diff
@@ -118,6 +118,9 @@ func (o *histogramOperator) processInputSeries(vectors []model.StepVector) ([]mo
 		for i, seriesID := range vector.SampleIDs {
 			outputSeries := o.outputIndex[seriesID]
 			outputSeriesID := outputSeries.outputID
+			if outputSeriesID >= len(o.seriesBuckets) {
```
Fixes the test case but I wonder if this is correct :thinking:

fpetkovski replied:
I think this is correct, but maybe for the wrong reasons :) Maybe there can be another cause (like a bug) that can lead to this.

GiedriusS commented:
Mhm, probably also need an adjustment here: https://github.com/thanos-community/promql-engine/blob/main/execution/function/histogram.go#L166.

fpetkovski commented:
Looks like we are hitting https://github.com/thanos-community/promql-engine/blob/44ee950ce4b04963844e81c0e78aa4f24f0f578d/execution/function/histogram.go#L170-L173

The value for `le` is empty so when we try to parse the value we get an error and skip the series 😢

GiedriusS commented:
I reworked this PR after digging into the code. I'm confident that this is the correct fix. There are buggy two cases here:
* Original PromQL engine returns NaN if there is only one bucket. Support for comparing NaNs is sorely needed because we've missed a bug :smile: 
* `outputIndex[]` has type `*histogramSeries` now. If `i` element is `nil` then it means that the original metric is malformed i.e. has no or invalid `le` label :smile: 

Tests pass locally except that `Diff:` is empty due to lack of support for comparing NaNs. Also, I've added logging to `os.Stderr` into the options so that we would get some useful output in case of a panic.

@fpetkovski PTAL once you have some time :pray:

fpetkovski reviewed (commented):
Thanks for the fix, this also looks good to me. Just a few questions on potentially redundant checks.

GiedriusS commented on execution/function/histogram.go:
```diff
@@ -127,14 +137,16 @@ func (o *histogramOperator) processInputSeries(vectors []model.StepVector) ([]mo
-				step.SampleIDs = append(step.SampleIDs, uint64(i))
-				step.Samples = append(step.Samples, math.NaN())
+			// We need at least 2 buckets to calculate a quantile.
+			if len(stepBuckets) < 2 {
```
Good point, `stepBuckets` will always be of length 1 or greater.

GiedriusS replied:
Actually, I've hit the condition `len(stepBuckets) == 0` somehow. It leads to the following crash:

```
runtime error: index out of range [-1]
...
github.com/thanos-community/promql-engine/execution/function.bucketQuantile(0x3feccccccccccccd, {0x0?, 0x0, 0x0})\n\t/ho
me/giedrius/go/pkg/mod/github.com/vinted/promql-engine@v0.0.0-20221205101229-9ef452546ade/execution/function/quantile.go:67 +0x3c9\ngithub.com/thanos-community/promql-engine/execution/function.(*histogra
mOperator).processInputSeries(0xc079ab7550, {0xc0ca9fad80, 0xa, 0xa})\n\t/home/giedrius/go/pkg/mod/github.com/vinted/promql-engine@v0.0.0-20221205101229-9ef452546ade/execution/function/histogram.go:145 +
0x725\ngithub.com/thanos-community/promql-engine/execution/function.(*histogramOperator).Next(0xc079ab7550, {0x2ba6f60, 0xc079a9b040})\n\t/home/giedrius/go/pkg/mod/github.com/vinted/promql-engine@v0.0.0-
20221205101229-9ef452546ade/execution/function/histogram.go:114
...
```

Trying to dig how this could happen.

fpetkovski reviewed (approved):
Thanks for the thorough work 👍
